scipy/signal/tests/test_signaltools.py

```diff
@@ -3521,194 +3921,227 @@ def test_initial_conditions_3d_axis1(self, dt):
         y2, z2 = sosfilt(sos, x[:, 5:, :], axis=axis, zi=z1)
 
         # y should equal yf, and z2 should equal zf.
-        y = np.concatenate((y1, y2), axis=axis)
-        assert_allclose_cast(y, yf, rtol=1e-10, atol=1e-13)
-        assert_allclose_cast(z2, zf, rtol=1e-10, atol=1e-13)
+        concat = array_namespace(x).concat
+        y = concat((y1, y2), axis=axis)
+        xp_assert_close(y, yf, rtol=1e-10, atol=1e-13)
+        xp_assert_close(z2, zf, rtol=1e-10, atol=1e-13)
 
         # let's try the "step" initial condition
         zi = sosfilt_zi(sos)
-        zi.shape = [nsections, 1, 2, 1]
+        zi = xp.reshape(zi, (nsections, 1, 2, 1))
         zi = zi * x[:, 0:1, :]
         y = sosfilt(sos, x, axis=axis, zi=zi)[0]
         # check it against the TF form
         b, a = zpk2tf(*zpk)
+        b, a = xp.asarray(b), xp.asarray(a)    # XXX while zpk2tf is numpy-only
+
         zi = lfilter_zi(b, a)
-        zi.shape = [1, zi.size, 1]
+        zi = xp.reshape(zi, (1, xp_size(zi), 1))
         zi = zi * x[:, 0:1, :]
         y_tf = lfilter(b, a, x, axis=axis, zi=zi)[0]
-        assert_allclose_cast(y, y_tf, rtol=1e-10, atol=1e-13)
+        xp_assert_close(y, y_tf, rtol=1e-10, atol=1e-13)
 
-    def test_bad_zi_shape(self, dt):
+    def test_bad_zi_shape(self, dt, xp):
+        dt = getattr(xp, dt)
         # The shape of zi is checked before using any values in the
         # arguments, so np.empty is fine for creating the arguments.
-        x = np.empty((3, 15, 3), dt)
-        sos = np.zeros((4, 6))
-        zi = np.empty((4, 3, 3, 2))  # Correct shape is (4, 3, 2, 3)
+        x = xp.empty((3, 15, 3), dtype=dt)
+        sos = xp.zeros((4, 6))
+        zi = xp.empty((4, 3, 3, 2))  # Correct shape is (4, 3, 2, 3)
         with pytest.raises(ValueError, match='should be all ones'):
             sosfilt(sos, x, zi=zi, axis=1)
         sos[:, 3] = 1.
         with pytest.raises(ValueError, match='Invalid zi shape'):
             sosfilt(sos, x, zi=zi, axis=1)
 
-    def test_sosfilt_zi(self, dt):
+    def test_sosfilt_zi(self, dt, xp):
+        dt = getattr(xp, dt)
         sos = signal.butter(6, 0.2, output='sos')
+        sos = xp.asarray(sos)   # XXX while butter is np-only
         zi = sosfilt_zi(sos)
 
-        y, zf = sosfilt(sos, np.ones(40, dt), zi=zi)
-        assert_allclose_cast(zf, zi, rtol=1e-13)
+        y, zf = sosfilt(sos, xp.ones(40, dtype=dt), zi=zi)
```

Comment:
Something interesting is happening with the test failure here via i.e., `python dev.py test -t scipy/signal/tests/test_signaltools.py::TestSOSFilt::test_sosfilt_zi -b cupy`

I noticed that `zf` in CuPy is the transpose of the result provided by NumPy and `torch`; i.e., this causes the above incantation to pass:

```diff
--- a/scipy/signal/tests/test_signaltools.py
+++ b/scipy/signal/tests/test_signaltools.py
@@ -3954,6 +3954,7 @@ class TestSOSFilt:
         with pytest.raises(ValueError, match='Invalid zi shape'):
             sosfilt(sos, x, zi=zi, axis=1)
 
+    @skip_xp_backends(cpu_only=True, exceptions=['cupy'])
     def test_sosfilt_zi(self, dt, xp):
         dt = getattr(xp, dt)
         sos = signal.butter(6, 0.2, output='sos')
@@ -3961,7 +3962,7 @@ class TestSOSFilt:
         zi = sosfilt_zi(sos)
 
         y, zf = sosfilt(sos, xp.ones(40, dtype=dt), zi=zi)
-        xp_assert_close(zf, zi, rtol=1e-13, check_dtype=False)
+        xp_assert_close(zf.T, zi, rtol=1e-13, check_dtype=False)
 
         # Expected steady state value of the step response of this filter:
         ss = xp.prod(xp.sum(sos[:, :3], axis=-1) / xp.sum(sos[:, 3:], axis=-1))
```

We obviously can't/shouldn't use a patch like that, filtered to CuPy case. At first glance, I can't imagine this would be an intentional change on the CuPy side based on the docs. I do see that the `zi` shape convention is a bit different between SciPy and CuPy, but upon closer inspection I wasn't entirely convinced that was the cause since `sosfilt_zi` should compensate by preserving the larger `zi` convention in CuPy anyway.